torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -391,11 +393,23 @@ def _init_param_attributes(self, p: Parameter) -> None:
                 )
             return
 
+        if hasattr(p, "_local_shard") and p.data != p._local_shard:  # type: ignore[attr-defined]
```

Comment:
This seems like it should be an error? `p.data` is not the shard anymore?

Reply:
it is possible that p.data is still pointing to sharded data, e..g, users convert the shard from fp16 to fp32 outside fsdp, but I agree that we should not support this use case, because users could do anything and we lose the check here